consensus/propagation/have_wants.go

```diff
@@ -63,58 +66,143 @@ func (blockProp *Reactor) handleHaves(peer p2p.ID, haves *proptypes.HaveParts) {
 		return
 	}
 
-	reqLimit := 1
-
-	// if enough requests have been made for the parts, don't request them.
-	for _, partIndex := range hc.GetTrueIndices() {
-		reqs := blockProp.countRequests(height, round, partIndex)
-		if len(reqs) >= reqLimit {
-			hc.SetIndex(partIndex, false)
-			// mark the part as fully requested.
-			fullReqs.SetIndex(partIndex, true)
+	for _, index := range hc.GetTrueIndices() {
+		select {
+		case <-blockProp.ctx.Done():
+		case p.requestChan <- request{height: height, round: round, index: uint32(index)}:
 		}
-		// don't request the part from this peer if we've already requested it
-		// from them.
-		for _, p := range reqs {
-			// p == peer means we have already requested the part from this peer.
-			if p == peer {
-				hc.SetIndex(partIndex, false)
+	}
+
+	blockProp.broadcastHaves(haves, peer, int(parts.Total()))
+}
+
+// perPeerPerBlockConcurrentRequestLimit returns the maximum number of requests for a given block by peer.
+// it helps avoid downloading the whole block from a single peer.
+func perPeerPerBlockConcurrentRequestLimit() int64 {
+	// TODO implement
+	return 300
+}
+
+func (blockProp *Reactor) wantsSendingRoutine(ps *PeerState) {
+	var requestsBA *bits.BitArray
+	height := int64(0)
+	round := int32(0)
+	periodicRequestsDelay := 10 * time.Millisecond
+	periodicRequestTicker := time.NewTicker(periodicRequestsDelay)
+	for {
+		select {
+		case <-blockProp.ctx.Done():
+			return
+		case <-periodicRequestTicker.C:
+			// periodically send the existing requests
+			if requestsBA != nil && !requestsBA.IsEmpty() {
+				blockProp.sendWant(ps, height, round, requestsBA)
+				requestsBA = nil
+			}
+		case req, ok := <-ps.requestChan:
+			if !ok {
+				return
+			}
+
+			// concurrent per-peer per-block request limit reached
+			if ps.requestCount.Load() >= perPeerPerBlockConcurrentRequestLimit() {
+				// we reached the limit of requests, we request what we can
+				if requestsBA != nil && !requestsBA.IsEmpty() {
+					blockProp.sendWant(ps, height, round, requestsBA)
+					requestsBA = nil
+				}
```

Comment:
are we sending the same wants to the same peer here potentially? if so, won't they send us that part each time we request it?

Reply:
can you explain what you mean by: won't they send us that part each time we request it?

Reply:
if we're not updating requestsBA after we send it, then we're asking for the same parts multiple times

I'm just trying to make sure that we're not doing that but its kinda difficult to track everything that touches it

Reply:
requestsBA is set to nil after sending the requests. Then, it's populated with new requests